src/odemis/dataio/hdf5.py

```diff
@@ -611,262 +611,106 @@ def _parse_physical_data(pdgroup, da):
-        try:
-            ds = pdgroup["TimeRange"]
-            timeRange = float(ds[i])
-            state = _h5svi_get_state(ds)
-            if state and state[i] == ST_INVALID:
-                raise ValueError
-            md[model.MD_STREAK_TIMERANGE] = timeRange
-        except (KeyError, IndexError, ValueError):
-            pass
-
-        try:
-            ds = pdgroup["MCPGain"]
-            MCPGain = float(ds[i])
-            state = _h5svi_get_state(ds)
-            if state and state[i] == ST_INVALID:
-                raise ValueError
-            md[model.MD_STREAK_MCPGAIN] = MCPGain
-        except (KeyError, IndexError, ValueError):
-            pass
-
-        try:
-            ds = pdgroup["StreakMode"]
-            streakMode = bool(ds[i])
-            state = _h5svi_get_state(ds)
-            if state and state[i] == ST_INVALID:
-                raise ValueError
-            md[model.MD_STREAK_MODE] = streakMode
-        except (KeyError, IndexError, ValueError):
-            pass
+            # MicroscopeMode helps us to find out the bandwidth of the wavelength
+            # and it's also a way to keep it stable, if saving the data again.
+            h_width = 1e-9  # 1 nm : default is to just almost keep the value
+            try:
+                mm = pdgroup["MicroscopeMode"][index]
+                if mm == MM_FLUORESCENCE:
+                    h_width = 10e-9  # 10 nm => narrow band
+                if mm == MM_TRANSMISSION:  # we set it for brightfield
+                    h_width = 100e-9  # 100 nm => large band
+            except Exception as ex:
+                raise ex  # TODO  does it now raise the exception below or do we also need a logging info here?
+                # TODO didn't know which data to use for testing
+
+            if MD_key is model.MD_IN_WL:
+                md[MD_key] = (float(ds[index]) - h_width, float(ds[index]) + h_width)  # in m
+            elif MD_key is model.MD_OUT_WL:
+                if isinstance(ds[index], basestring):
+                    md[model.MD_OUT_WL] = ds[index]
+                elif len(ds.shape) == 1:  # Only one value per channel
+                    ewl = float(ds[index])  # in m
+                    # In files saved with Odemis 2.2, MD_OUT_WL could be saved with
+                    # more precision in C scale (now explicitly saved as tuple here)
+                    if MD_key not in md:
+                        md[MD_key] = (ewl - h_width, ewl + h_width)
+                else:  # full band for each channel
+                    md[MD_key] = tuple(ds[index])
 
-        try:
-            ds = pdgroup["TriggerDelay"]
-            triggerDelay = float(ds[i])
-            state = _h5svi_get_state(ds)
-            if state and state[i] == ST_INVALID:
-                raise ValueError
-            md[model.MD_TRIGGER_DELAY] = triggerDelay
-        except (KeyError, IndexError, ValueError):
-            pass
+        else:
+            data_type = type(ds[index])
```

Comment:
Instead of using the data type of the entry as it's stored in the file, I'd suggest to get as argument the `converter`, which is a function which takes the original data and returns the value as it should be stored in the metadata. So the most typical "converters" will be `float`,`str`...

There are two reasons for doing so. First, it'd make it clear inside the code which metadata is which type. Second, and most important, is that it wouldn't depend on the input file. This is a general "safety" principle in software: do not rely on the user input to be correct. With this current version, if for some reason the HDF5 has "exposure time" as a string, the MD_EXP_TIME will be set as a string. So opening the file will work... and then random weird things will happen in the GUI. IOW, it's better to be very defensive for every input that you don't absolutely control.

Reply:
yes good idea. Was looking for something like that, as I also didn't like my implementation that much.